doc/source/data/data-internals.rst

```diff
@@ -92,6 +92,44 @@ To change the size at which Ray Data splits blocks, configure
 Ray Data can’t split rows. So, if your dataset contains large rows (for example, large
 images), then Ray Data can’t bound the block size.
 
+
+Shuffle Algorithms
+------------------
+
+Ray Data implements two main shuffle algorithms that power different operations:
+
+.. _hash-shuffle:
+
+Hash-shuffling
+~~~~~~~~~~~~~
+
+.. note:: Hash-shuffling is available in Ray 2.46
+
+Hash-shuffling is a classical hash-partitioning based shuffling where:
+
+1. **Partition phase:** rows in every block are hash-partitioned based on values in the *key columns* into a specified number of partitions, following a simple residual formula of ``hash(key-values) % N`` (used in hash-tables and pretty much everywhere).
+2. **Push phase:** partition's shards from individual blocks are then pushed into corresponding aggregating actors (called ``HashShuffleAggregator``s) handling respective partitions.
+3. **Reduce phase:** aggregators combine received individual partition's shards back into blocks optionally applying additional transformations before producing the resulting blocks.
```

Comment:
Does (2) refer to "submit", and (3) refers to "finalize"? Or does (3) include both? 
I think we can explain that the actual "reduce" operation includes 2 phases: a streaming partial reduce (the "submit") and a finalization.

Reply:
This refers to both in more abstract terms. I wanted to keep this on a higher level to avoid burying it in technical details. 

Goal of this section is just to educate users how this works internally, rather than provide code pointers.

Reply:
It's fine to not talk about many technical details here. 
I think the main confusion is that this may make readers think that each of these phases is blocking. So I'd like to point out that the hash-partition and reduce ops are overlapped, which is a key difference from the sort shuffling.